compiler/rustc_parse/src/parser/item.rs

```diff
@@ -670,14 +670,44 @@ impl<'a> Parser<'a> {
             }
             match parse_item(self) {
                 Ok(None) => {
+                    let is_unnecessary_semicolon = !items.is_empty()
+                        // When the close delim is `)` in a case like the following, `token.kind` is expected to be `token::CloseDelim(Delimiter::Parenthesis)`,
+                        // but the actual `token.kind` is `token::CloseDelim(Delimiter::Bracket)`.
+                        // This is because the `token.kind` of the close delim is treated as the same as
+                        // that of the open delim in `TokenTreesReader::parse_token_tree`, even if the delimiters of them are different.
+                        // Therefore, `token.kind` should not be compared here.
+                        //
+                        // issue-60075.rs
+                        // ```
+                        // trait T {
+                        //     fn qux() -> Option<usize> {
+                        //         let _ = if true {
+                        //         });
+                        //          ^ this close delim
+                        //         Some(4)
+                        //     }
+                        // ```
+                        && self
+                            .span_to_snippet(self.prev_token.span)
+                            .map_or(false, |snippet| snippet == "}")
+                        && self.token.kind == token::Semi;
+                    let semicolon_span = self.token.span;
                     // We have to bail or we'll potentially never make progress.
                     let non_item_span = self.token.span;
                     self.consume_block(Delimiter::Brace, ConsumeClosingDelim::Yes);
-                    self.struct_span_err(non_item_span, "non-item in item list")
-                        .span_label(open_brace_span, "item list starts here")
+                    let mut err = self.struct_span_err(non_item_span, "non-item in item list");
+                    err.span_label(open_brace_span, "item list starts here")
                         .span_label(non_item_span, "non-item starts here")
-                        .span_label(self.prev_token.span, "item list ends here")
-                        .emit();
+                        .span_label(self.prev_token.span, "item list ends here");
+                    if is_unnecessary_semicolon {
+                        err.span_suggestion_verbose(
+                            semicolon_span,
```

Comment:
semicolon_span seems does not take effect..

Reply:
> semicolon_span does not take effect

what does this mean?

Reply:
I mean `help: consider removing this semicolon` will be placed at this line,
https://github.com/rust-lang/rust/blob/48c0341a7081bb74bb4635529a94298b9be4b459/src/test/ui/parser/suggest-removing-semicolon-after-impl-trait-items.stderr#L7

The code used `semicolon_span`, but it's value seems wrong?

Reply:
The span is correct, because the `// run-rustfix` test correctly fixes the code. The rendering is wrong, since "consider removing this semicolon" is put to the bottom of the error message.

This is a rendering bug, possibly due to the usage of `span_suggestion_verbose`, but I think @TaKO8Ki can investigate.